posthog/models/historical_version.py

```diff
@@ -0,0 +1,112 @@
+import datetime
+import json
+
+from django.db import models
+from django.utils import timezone
+
+from posthog.models import FeatureFlag
+from posthog.models.utils import UUIDModel
+
+
+class HistoricalVersionJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, models.base.ModelState):
+            return None
+        if isinstance(obj, FeatureFlag):
+            return obj.__dict__
+        if isinstance(obj, datetime.datetime):
+            return obj.isoformat()
+
+        return json.JSONEncoder.default(self, obj)
+
+
+class HistoricalVersion(UUIDModel):
+    """
+    We don't store foreign key references to the item being versioned
+    because the referenced model will change or be deleted.
+
+    The history log should hold the state at the time it is written not the time it is read
```

Comment:
Not sure I understand this line

Reply:
Equivalent to pass by reference or value. If I hold a reference to X someone can change X and when I generate the log before and after that change the log is different but should be immutable

You need the state of the item as it was each time the log is written to. Not the state of the item as it is when the log is read from

You _could_ store the changes instead of the state and then you wouldn't care, but changing how you calculate changes would then also include changes over the data. This way if we alter how we calculate changes the log reflects the new logic when it is next read

Reply:
This was my way of saying "the line is confusing or misleading, can you clear it up". E.g. why "should hold", not "actually does hold"? Why would I even assume that when I read a history log I get current data (as it is when I read it...?)? What is even "it" here, the log or the item?

I'm still not sure I understand this after the expanded explanation 😅